subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/AmbiguousTransformException.java

```diff
@@ -33,13 +37,31 @@ public AmbiguousTransformException(String producerDisplayName, AttributeContaine
 
     private static String format(String producerDisplayName, AttributeContainerInternal requested, List<Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant>> candidates) {
         TreeFormatter formatter = new TreeFormatter();
-        formatter.node("Found multiple transforms that can produce a variant of " + producerDisplayName + " for consumer attributes");
+        formatter.node("Found multiple transforms that can produce a variant of " + producerDisplayName + " with requested attributes");
         formatAttributes(formatter, requested);
         formatter.node("Found the following transforms");
+        Map<ResolvedVariant, List<ConsumerVariantMatchResult.ConsumerVariant>> variantToTransforms = candidates.stream()
+            .collect(Collectors.toMap(candidate -> candidate.getLeft(),
+                    candidate -> Lists.newArrayList(candidate.getRight()),
+                    (List<ConsumerVariantMatchResult.ConsumerVariant> orig, List<ConsumerVariantMatchResult.ConsumerVariant> add) -> {
+                        orig.addAll(add);
+                        return orig;
+                    },
+                    LinkedHashMap::new));
         formatter.startChildren();
-        for (Pair<ResolvedVariant, ConsumerVariantMatchResult.ConsumerVariant> candidate : candidates) {
-            formatter.node("Transform from " + candidate.getLeft().asDescribable().getDisplayName());
-            formatAttributes(formatter, candidate.getLeft().getAttributes());
+        for (Map.Entry<ResolvedVariant, List<ConsumerVariantMatchResult.ConsumerVariant>> entry : variantToTransforms.entrySet()) {
+            formatter.node("From variant '" + entry.getKey().asDescribable().getDisplayName() + "'");
+            formatter.startChildren();
+            formatter.node("With source attributes");
+            formatAttributes(formatter, entry.getKey().getAttributes());
+            formatter.node("Candidate transformation(s)");
+            formatter.startChildren();
+            for (ConsumerVariantMatchResult.ConsumerVariant transform : entry.getValue()) {
+                formatter.node("Transformation '" + transform.transformation.getDisplayName() + "' producing attributes:");
```

Comment:
So we would have one entry for a chained transformation, right?

Should we call theses things `Transform(s)` when reporting them? `Transformation` would be something new exposed on the public API.

Reply:
Indeed, will update